aarch64/kernel/nano_core/main.rs

```diff
@@ -47,6 +50,17 @@ fn main(
     let mmap_size = boot_svc.memory_map_size();
     let mut mmap = vec![0; mmap_size.map_size + safety * mmap_size.entry_size];
 
+    let mut mapped_regions = 0;
+    {
+        let (_, layout) = boot_svc.memory_map(&mut mmap).unwrap();
+        for descriptor in layout {
+            if descriptor.ty != MemoryType::CONVENTIONAL {
+                mapped_regions += 1;
+            }
+        }
+    }
+    let mut mapped_regions = Vec::with_capacity(mapped_regions + safety);
```

Comment:
what is this for? I'm a bit confused why we're iterating over available memory regions from UEFI multiple times

Reply:
The code counts the number of mapped regions so that it knows precisely how many slots to allocate in `Vec::with_capacity`

Reply:
`layout` is only usable once (owned Iterator implementer) so the code iterates on it only, and then later when boot services are exited we have the pre-allocated vec where we can push mapped regions

Reply:
I see, sounds good. One more question -- can we continue to safely use memory that was allocated by UEFI's heap allocator _after_ we've exited boot services? That is, `mapped_regions` belongs in that heap allocator's memory, which no longer exists after we exit boot services, right?

Reply:
It's forbidden to call allocator functions but the memory is still mapped after boot services are exited, and the regions are present in the returned layout that we use to build the new page table.

Reply:
Ok, so once that heap item (the `Vec`) gets dropped it will call heap deallocation routines, so we must ensure that happens before exiting boot services, right? If not, won't undefined behavior occur by calling dealloc in a now-unusable heap allocator?

Reply:
I simplified what's done in `nano_core` in the last commits